Enable config file based configuration

MrEcco opened the pull request:
This is a little rethink of how to configure. Since the previous implementation disallows to use the compiled binaries in releases, I have prepared a more flexible version. However, the legacy behavior (without configuration file) remains as is (proven by unit tests).

## List of changes

- Enable `-config` flag. If no such flag provided, then remain the original behavior
- Partial refactoring of the initialization process
- Fix small issue what prevented multiple hypervisors connecting to the controlling socket (e.g. now you can attach more than 1 qemu VM, and they can ping each other)

Most of all changes are tested, see unit tests.

## Reasons

I'm trying to enable multi-VM local lab which should supports different host OS, guest OS, guest arches. With this tool I have resolved the significant problem with the proper VM networking (at least for MacOS). However, I must fix a few issues, here's the PR.

Pealse accept it and release the new version of gvproxy binary. I would appreciate.

lstocchi reviewed (commented):
Looking at the code LGTM, need to do some test.

Adding @cfergeau and @evidolob as reviewers so they give it a look as well

MrEcco commented on cmd/gvproxy/config.go:
```diff
@@ -0,0 +1,472 @@
+	}
+
+	if InDebugMode() {
+		config.Stack.Debug = true
```
This is used in the Stack definition, outside the changes in this PR.

Original solution: https://github.com/containers/gvisor-tap-vsock/blob/main/cmd/gvproxy/main.go#L217
The structure: https://github.com/containers/gvisor-tap-vsock/blob/main/pkg/types/configuration.go#L10

Sure, I can set it to something like `config.Stack.Debug = InDebugMode()` to bring some single responsibility here :) But it may break the original behavior.

MrEcco commented on cmd/gvproxy/config.go:
```diff
@@ -0,0 +1,472 @@
+			return config, errors.New("ssh-port value must be between 1024 and 65535")
+		}
+
+		if config.Stack.CaptureFile == "" {
```
That's fair. Thanks 😄

MrEcco commented on cmd/gvproxy/config.go:
```diff
@@ -0,0 +1,472 @@
+					},
+					{
+						Name: host,
+						IP:   net.ParseIP(hostIP),
```
Originally, it has been hardcoded: https://github.com/containers/gvisor-tap-vsock/blob/main/cmd/gvproxy/main.go#L52-L58

If I understand right, originally maintainers just use/debug/etc it with changes in these settings and do `go run . -atata`. I wanted to remain this ability.

Please notify if you want me to set it to luaddr anyway 😄

MrEcco commented:
Any news on it?

cfergeau commented:
After the addition of the DCO, there is no longer a commit log to your change :-/ The initial description in the PR should be good as a commit log.
Can you also rebase this (and not merge) on top of `main`? There will a conflict related to https://github.com/containers/gvisor-tap-vsock/commit/4bbb832355b8d141a61694e322ce5cbb191d28c9 , I’m a bit worried of introducing a regression if I try to fix it myself :-/

MrEcco commented on cmd/gvproxy/config.go:
```diff
@@ -0,0 +1,470 @@
+	"gopkg.in/yaml.v3"
+)
+
+// Legacy hardcode
```
I've got rid of all mentions of "legacy" in favor to "the default".

MrEcco commented on cmd/gvproxy/config.go:
```diff
@@ -0,0 +1,470 @@
+	if config.LogFile != "" {
+		lf, err := os.Create(config.LogFile)
+		if err != nil {
+			fmt.Printf("unable to open log file %s, exiting...\n", config.LogFile)
```
It was just a copy-paste from the previous code. Yes, it should be `return error` instead of `os.Exit`.

In the separated commit, I've moved it to the start of the same function.

MrEcco commented on cmd/gvproxy/config.go:
```diff
@@ -0,0 +1,470 @@
+
+	if args.config != "" {
+		if slices.Contains(os.Args, "-ssh-port") || slices.Contains(os.Args, "--ssh-port") {
+			log.Warningf("CLI argument \"-ssh-port\" is unavailable with config file. You need to add \"127.0.0.1:%d: 192.168.127.2:22\" entry into .stack.forwards in \"\" instead", args.sshPort)
```
I want to remain CLI flags as is as much as possible, because I'm not a maintainer here, you know. If you just say that I can reimplement the initialization from scratch, I would do it :) But it may disturb maintainers, no one likes when someone changes the code they already got used to.

If you propose me to reimplement configuration of it without default behavior with hardcoded configuration, I would do, with bells, tests, and whistles. Maybe with some portion of integration testing. Please answer in thread :)

MrEcco replied:
BTW, cli args and config files complement each other. You can find the proof of it in tests.

cfergeau replied:
> BTW, CLI args and config file complement each other. You can find the proof of it in tests.

I’ve seen that you can override some parts of the config with commandline args, but looking at the tests, I don’t really see tests for stuff like:
```
--config config.yaml --listen ... --forward... --ssh-port 1234
```
I think the code in the PR has kind support for matching both, but I was asking if you have a requirement for this? we could decide `--config` and other network-related args can’t be used together.

MrEcco commented on cmd/gvproxy/config.go:
```diff
@@ -0,0 +1,470 @@
+const (
+	// 	gatewayIP   = "192.168.127.1"
+	sshHostPort = "192.168.127.2:22"
+	hostIP      = "192.168.127.254"
```
Yes, but please look at the original code. It seems like the maintainers used this code with these hardcoded variables and calls like `go run main.go`, without flags. I don't know if I can remove this feature.

cfergeau replied:
I don’t expect many people to use `go run main.go` directly, at the very least you need to specify one of the `listen` args.
The main user of `gvproxy` that I’m aware of is podman-machine,  mainly for use on macos and windows, but on linux it also starts gvproxy. Its commandline (on macos for example) is:
```
/opt/podman/bin/gvproxy -mtu 1500 -ssh-port 64450 -listen-vfkit unixgram:///var/folders/l0/rh4v8_j54k37h2w320__7d1h0000gn/T/podman/podman-machine-default-gvproxy.sock -forward-sock /var/folders/l0/rh4v8_j54k37h2w320__7d1h0000gn/T/podman/podman-machine-default-api.sock -forward-dest /run/user/501/podman/podman.sock -forward-user core -forward-identity /Users/teuf/.local/share/containers/podman/machine/machine -pid-file /var/folders/l0/rh4v8_j54k37h2w320__7d1h0000gn/T/podman/gvproxy.pid -log-file /var/folders/l0/rh4v8_j54k37h2w320__7d1h0000gn/T/podman/gvproxy.log
```

Imo the main reason for these hardcoded values is because it was easier this way rather than writing the code you added, where you specify the subnet you want, and everything is inferred from it.

MrEcco commented on cmd/gvproxy/config.go:
```diff
@@ -0,0 +1,470 @@
+		} else {
+			os.MkdirAll("/var/run/gvproxy", os.ModePerm)
+			return []string{"unix:///var/run/gvproxy/default.sock"}
+		}
```
on MacOS:
- `ls -al /var/run` is `drwxrwxr-x 0 root daemon ...`
- `ls -al /opt/homebrew/var/run` is `drwxr-xr-x 0 username admin ...`

1. No, `/var/run` requires root permissions
2. However, it's hardly possible that someone who uses this tool has no `homebrew` installed
3. I'm going to distribute this binary with `brew`, in my own tap if RH doesn't
4. For sure, it's not absolutely necessary, just a preferable thing

cfergeau replied:
gvproxy is currently part of the podman formula https://github.com/Homebrew/homebrew-core/blob/f56f62cc378ca260be28389b450d1c370d942e21/Formula/p/podman.rb#L58-L63 , so in a way it’s already in brew. You’ll probably want to split it to make it easier to install.

> However, it's hardly possible that someone who uses this tool has no homebrew installed

If you install https://github.com/containers/podman/releases/download/v5.4.1/podman-installer-macos-arm64.pkg , you get gvproxy without brew.

MrEcco commented on cmd/gvproxy/config.go:
```diff
@@ -0,0 +1,470 @@
+		for i := 0; i < bits/16; i++ {
+			m[i] = m1[i]
+			m[i+(bits/16)] = m2[i]
+		}
```
It's not broadcast IP, but I've got your point, I'll try to adopt this shorter code.

MrEcco replied:
I found even more beautiful solution: just IP bytes + inverted mask bytes and -1 for the last byte produce the correct result: https://github.com/containers/gvisor-tap-vsock/pull/484/commits/e82d43266bb32ff9b227da46ad17941a496224d1

MrEcco commented on cmd/gvproxy/config.go:
```diff
@@ -0,0 +1,470 @@
+		if args.config != "" {
+			config.Listen = getDefaultListen(runtime.GOOS)
+		}
+	}
```
Yes, maybe it's better to remove this thing.

MrEcco commented on cmd/gvproxy/main.go:
```diff
@@ -362,137 +175,164 @@ func run(ctx context.Context, g *errgroup.Group, configuration *types.Configurat
+					g.Go(func() error {
+						return vn.AcceptQemu(ctx, conn)
+					})
+				}
```
It's pretty small change which enabled multiple Qemu instances in a single virtual network. Isn't it as it's supposed to be? I actively use it for testing connectivity between such Qemu instances.

MrEcco replied:
https://github.com/containers/gvisor-tap-vsock/pull/484#discussion_r2010899562

MrEcco commented on cmd/gvproxy/main.go:
```diff
@@ -362,137 +175,164 @@ func run(ctx context.Context, g *errgroup.Group, configuration *types.Configurat
-					return vn.AcceptVpnKit(conn)
-				})
 			}
-			return nil
```
Moved the code related to multiaccept fix into separated commit in this PR: https://github.com/containers/gvisor-tap-vsock/pull/484/commits/7b8db37632f41c1907511bc017ba56d08530f665

MrEcco commented on cmd/gvproxy/main.go:
```diff
@@ -362,137 +175,164 @@ func run(ctx context.Context, g *errgroup.Group, configuration *types.Configurat
+					})
+				}
 			}
-			return vn.AcceptBess(ctx, conn)
```
https://github.com/containers/gvisor-tap-vsock/pull/484#discussion_r2010899562

MrEcco commented on cmd/gvproxy/main.go:
```diff
@@ -362,137 +175,164 @@ func run(ctx context.Context, g *errgroup.Group, configuration *types.Configurat
+					g.Go(func() error {
+						return vn.AcceptVfkit(ctx, vfkitConn)
+					})
+				}
```
https://github.com/containers/gvisor-tap-vsock/pull/484#discussion_r2010899562

MrEcco commented on cmd/gvproxy/config.go:
```diff
@@ -0,0 +1,470 @@
+			if err := lf.Close(); err != nil {
+				fmt.Printf("unable to close log-file: %q\n", err)
+			}
+		}()
```
Fixed here: https://github.com/containers/gvisor-tap-vsock/pull/484/commits/f5a0e2eb460768f38b287230eddd7f509e69768f

MrEcco commented:
@cfergeau Thanks a lot for your review! I have rebased it properly (I'll never use GH UI button for it :) ), fixed all findings, cleaned up the linter's output (only for files in the scope of this PR), and have done my best with atomic commits. Please reply in the remaining threads, I don't really sure what to do there, or just resolve.

I hope it's done here :)

cfergeau commented:
I’ve rebased this and squashed the commits which go together in https://github.com/cfergeau/gvisor-tap-vsock/tree/MrEcco
The modified version works fine with podman-machine and macadam on a mac. I need to take a closer look at the "multiaccept" changes though.

MrEcco commented:
Sorry, I forgot about this PR. Months ago, I have reimplemented your idea with appropriate configuration (or just with my own understanding of "appropriate", for my case) and CLI interface, here: https://github.com/circumspectlabs/gvswitch. This little project is a well tested by my students, they use it with this Ansible code which can start multiple VMs with a shared private network: https://github.com/circumspectlabs/kubernetes-the-mindful-way/blob/main/code/molecule/default/create.yml#L255-L272.

You can take the CLI and config implementation from this repo. It works with 6 VMs (haven't succeeded to start more on my laptop). Haven't yet tested on Windows and Linux.
